tracer/src/Datadog.Trace/DataStreamsMonitoring/DataStreamsManager.cs

```diff
@@ -0,0 +1,135 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DataStreamsMonitoring.Hashes;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Headers;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Vendors.Serilog.Events;
+
+namespace Datadog.Trace.DataStreamsMonitoring;
+
+/// <summary>
+/// Manages all the data streams monitoring behaviour
+/// </summary>
+internal class DataStreamsManager
+{
+    private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<DataStreamsManager>();
+    private readonly NodeHashBase _nodeHashBase;
+    private bool _isEnabled;
+
+    public DataStreamsManager(
+        bool enabled,
+        string env,
+        string defaultServiceName)
+    {
+        _isEnabled = enabled;
+        // We don't yet support primary tag in .NET yet
+        _nodeHashBase = HashHelper.CalculateNodeHashBase(defaultServiceName, env, primaryTag: null);
+    }
+
+    public bool IsEnabled => Volatile.Read(ref _isEnabled);
+
+    public static DataStreamsManager Create(
+        ImmutableTracerSettings settings,
+        string defaultServiceName)
+        => new(settings.IsDataStreamsMonitoringEnabled, settings.Environment, defaultServiceName);
+
+    public Task DisposeAsync()
+    {
+        Volatile.Write(ref _isEnabled, false);
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Trys to extract a <see cref="PathwayContext"/>, from the provided <paramref name="headers"/>
+    /// If data streams is disabled, or no pathway is present, returns null.
+    /// </summary>
+    public PathwayContext? ExtractPathwayContext<TCarrier>(TCarrier headers)
+        where TCarrier : IBinaryHeadersCollection
+        => IsEnabled ? DataStreamsContextPropagator.Instance.Extract(headers) : null;
+
+    /// <summary>
+    /// Injects a <see cref="PathwayContext"/> into headers
+    /// </summary>
+    /// <param name="context">The pathway context to inject</param>
+    /// <param name="headers">The header collection to inject the headers into</param>
+    public void InjectPathwayContext<TCarrier>(PathwayContext? context, TCarrier headers)
+        where TCarrier : IBinaryHeadersCollection
+    {
+        if (!IsEnabled)
+        {
+            return;
+        }
+
+        if (context is not null)
+        {
+            DataStreamsContextPropagator.Instance.Inject(context.Value, headers);
+            return;
+        }
+
+        // This shouldn't happen normally, as you should call SetCheckpoint before calling InjectPathwayContext
+        // But if data streams was disabled, you call SetCheckpoint, and then data streams is enabled
+        // you will hit this code path
+        System.Diagnostics.Debug.Fail("Attempted to inject null pathway context");
```

Comment:
Should this `System.Diagnostics.Debug.Fail("Attempted to inject null pathway context");` be here?

Reply:
oops, no, thanks!